Refactor kir::ForLoop

naoyam opened the pull request:
Adds start, stop and step values to `kir::ForLoop`. Extracted from #809.

Commit:
Set start, stop and step in kir::ForLoop

Commit:
cleanup

Commit:
Validate allocations with unrestricted for-loops

Commit:
cleanup and validation

csarofeen commented:
Can you provide some more context on `unrestricted_parallel_extent_` and `replacement_map_`?

naoyam commented:
> Can you provide some more context on `unrestricted_parallel_extent_` and `replacement_map_`?

I removed `unrestricted_parallel_extent_` as it's not necessary at least for the shift PR. 

The intent is to support a case where the loop extent is larger than the number of threads. Currently, we assume the number of threads is always greater or equal to the loop extent, so it is always safe to not generate the for-loop. `unrestricted_parallel_extent_` was used to indicate whether that assumption may not hold. I simplified that part to just check whether the loop end is equal to the IterDomain extent. If that check is false, it conservatively assumes it is not safe to skip generating the for-loop.

https://github.com/csarofeen/pytorch/blob/forloop-refactor/torch/csrc/jit/codegen/cuda/codegen.cpp#L909-L919

`replacement_map_` is used when the for-loop is not generated so that uses of the loop index are replaced with the loop start value. For example, by default, here's the loop template:

```
for (int x = threadIdx.x; x < stop; x += blockDim.x) {
   do_some_comp(x);
}
```

When it's safe, we generate:

```
do_some_comp(threadIdx.x);
```

Commit:
Add comments

Commit:
clang-tidy

Commit:
trigger tests

csarofeen reviewed (approved):
LGTM thanks!
